Expand Persona links/locking to characters and allowing more control on which persona to choose/load

Wolfsblvt opened the pull request:
I have added the Copilot summary below. It's not too bad to be honest. Still I wanna explain the reasoning why I did this, and what was added in my own words. (Though commit messages explain most of everything)

The base idea was #3300, and likely many other issues or mentions that asked for better control on which Persona to be used.

What this PR does is the following.

> [!IMPORTANT]
> This PR will **remove** the functionality of user avatar/usernames that are not a real persona.
> In the future, all user-based interactions will be done via a persona.
> Existing username without persona will be migrated 

### Improvements
- Expanded the Persona Panel with headings, splitting off what persona-specific settings are and which are global settings
- Define new types to lock a persona: lock to **chat**, to **character** and as **default** persona
- Add toggle buttons in the Persona Panel to quickly toggle those lock states, and see which state they currently have for the current context
- Save character persona locks/connections to the persona objects
- Show characters this persona is locked to as an avatar list inside the persona panel
- Characters in that list will be clickable, so you can directly jump to a char and their last chat from a persona
- Update persona auto selection on chat open to load and respect all different kinds of persona locks
- Add popup if multiple personas are connected to the character, asking which persona to choose (highlighting personas used in this chat)
- Add new global settings for personas:
  - "Auto-lock to chat" whenever a new persona is selected - no matter if manually of via auto selection - saving it to the chat file
  - "Allow multiple connections" to make it possible to link multiple personas to a char, so that a popup appears to choose which persona you want, every time you open a new chat with that persona
- Add button to char panel that shows a popup for existing persona connections. Allows select and removal.
- Refactor the already existing global setting for "show persona toasts"
- Update/add `/persona-lock` slash command to allow specifying the lock type, and also returning the current lock state, if no state is provided (in accordance to other similar slash commands)
- Keep but deprecate the old `/lock` slash command that should go along with the new argument syntax in the future
- Reworked several toasts to be more clear on persona selection, changes and stuff. All of them only appear if persona toasts are enabled in the global persona settings.
- Moved persona list buttons (edit, duplicate, etc) to the right panel as current persona buttons - to be consistent with char list and for better usability
- Add connection states of all personas to the persona list as slidable "tag"
- Improve functionality to rename personas
- Remove functionality of non-persona username and avatar

### Technical changes
- Add utility function to print persona avatar lists
- Refactor all existing blocks that used `chid` or `grid` as an actual HTML attribute. It is now an actual data attribute.
- Allow multiline avatar display in places where it makes sense, via new CSS class
- Create common style template for hover-able images buttons/tags where the label only appears on hover
- Changed all persona backend calls from Ajax calls to fetch
- Refactor and move functions and slash commands to `persona.js`

![image](https://github.com/user-attachments/assets/f9af37b2-5bcf-4af3-8999-bc37f6137529)
![image](https://github.com/user-attachments/assets/26f983ef-991c-47ae-bc53-bc0649cb08e0)
![image](https://github.com/user-attachments/assets/a3a0a406-6246-4d44-bedd-7c47091402fd)
![image](https://github.com/user-attachments/assets/eecd40f3-b946-494e-866c-408cd37ac04b)
![image](https://github.com/user-attachments/assets/a0fbca7e-41c5-42ae-b7e8-cf8769775e5b)


## Copilot Summary

This pull request includes significant updates to the user interface and functionality related to persona management and character selection. The changes involve CSS adjustments, HTML restructuring, and new JavaScript functions to enhance user experience and maintainability.

### User Interface Enhancements:

* [`public/css/toggle-dependent.css`](diffhunk://#diff-1382de9c4d62c2a0b9b0aec179d52629d2180587d4b4d465f86cc26f728739a9L99-L103): Added new styles for multi-line avatars and removed unused avatar button styles. [[1]](diffhunk://#diff-1382de9c4d62c2a0b9b0aec179d52629d2180587d4b4d465f86cc26f728739a9L99-L103) [[2]](diffhunk://#diff-1382de9c4d62c2a0b9b0aec179d52629d2180587d4b4d465f86cc26f728739a9R229-R238)

### HTML Structure Improvements:

* [`public/index.html`](diffhunk://#diff-c2cc24bc9001b11b6add48a4cd8f893d5d6c6e4d1bd254158bd14ab997f552cdL4987-R4987): Reorganized the persona management section, added new buttons for persona controls, and enhanced the layout for better user interaction. Introduced new sections for persona connections and global settings. [[1]](diffhunk://#diff-c2cc24bc9001b11b6add48a4cd8f893d5d6c6e4d1bd254158bd14ab997f552cdL4987-R4987) [[2]](diffhunk://#diff-c2cc24bc9001b11b6add48a4cd8f893d5d6c6e4d1bd254158bd14ab997f552cdL5010-R5034) [[3]](diffhunk://#diff-c2cc24bc9001b11b6add48a4cd8f893d5d6c6e4d1bd254158bd14ab997f552cdR5057-R5080) [[4]](diffhunk://#diff-c2cc24bc9001b11b6add48a4cd8f893d5d6c6e4d1bd254158bd14ab997f552cdR5089-R5105) [[5]](diffhunk://#diff-c2cc24bc9001b11b6add48a4cd8f893d5d6c6e4d1bd254158bd14ab997f552cdL5087-R5131) [[6]](diffhunk://#diff-c2cc24bc9001b11b6add48a4cd8f893d5d6c6e4d1bd254158bd14ab997f552cdL5125-R5191) [[7]](diffhunk://#diff-c2cc24bc9001b11b6add48a4cd8f893d5d6c6e4d1bd254158bd14ab997f552cdL5483-R5541)

### JavaScript Functionality Updates:

* [`public/script.js`](diffhunk://#diff-4a4ec86e4c05b413e9b128d0f702c797e4a2afd8be513301c019824560d0c8e6R238-R242): Added new functions to handle persona connections, updated existing functions to support new persona management features, and improved character selection logic. [[1]](diffhunk://#diff-4a4ec86e4c05b413e9b128d0f702c797e4a2afd8be513301c019824560d0c8e6R238-R242) [[2]](diffhunk://#diff-4a4ec86e4c05b413e9b128d0f702c797e4a2afd8be513301c019824560d0c8e6R1370-R1377) [[3]](diffhunk://#diff-4a4ec86e4c05b413e9b128d0f702c797e4a2afd8be513301c019824560d0c8e6L1438-R1451) [[4]](diffhunk://#diff-4a4ec86e4c05b413e9b128d0f702c797e4a2afd8be513301c019824560d0c8e6R1577) [[5]](diffhunk://#diff-4a4ec86e4c05b413e9b128d0f702c797e4a2afd8be513301c019824560d0c8e6L6298-R6312)

<!-- Put X in the box below to confirm -->

## Checklist:

- [x] I have read the [Contributing guidelines](https://www.youtube.com/watch?v=vDWlhQjGjoc).

Wolfsblvt committed:
Refactor persona panel with headings

Wolfsblvt committed:
Reworked persona lock buttons + lock to character

- Slight redesign of the persona panel right side, add headers
- Add buttons for all three different lock states (chat, character, default)
- Update lock and toggle lock functions to accept toggle type
- Write persona locking to chat functionality (saves to persona description object)

Wolfsblvt committed:
Show connected characters in persona panel

- Print list of connected characters in persona panel
- Fix smaller bugs with missing connections in persona description

Wolfsblvt committed:
Add/expand persona auto selection from char / chat

- Persona selection on chat based on three levels:
  1. Chat Locking
  2. Char connected personas
  3. Default persona
- Add popup if multiple personas are connected to char
- Add utility function to print persona avatar lists

Wolfsblvt committed:
Persona settings for auto-lock & multi connection

- Add setting to auto-lock persona to chat. Triggered on chat changed (creation/loaded) and on persona selection.
- Add setting to allow multiple persona<->char connections, which then triggers the popup to select one
- If no multi connections allowed, the connect toggle will remove all existing connections from personas to that character

Wolfsblvt committed:
Update persona toasts

Wolfsblvt committed:
Merge branch 'staging' into persona-improvements

Wolfsblvt committed:
Refactor chid/grid attributes to data attributes

- We don't believe in imaginary HTML attributes that we make up, right?

Wolfsblvt committed:
Add css class to allow multiple avatar lists

Wolfsblvt committed:
Expand /persona-lock command for type

- Add new main alias for /lock, renamed to /persona-lock
- Allow no state to be provided to return the current lock state
- Deprecate the old usage of /lock without state, without breaking it

Wolfsblvt committed:
Fix one persona connected warning

Cohee1207 commented on public/scripts/world-info.js:
```diff
@@ -5149,7 +5149,7 @@ jQuery(() => {
 
     $('#world_button').on('click', async function (event) {
-        const chid = $('#set_character_world').data('chid');
+        const chid = $('#set_character_world').data('data-chid');
```
This probably has to be `attr` instead of `data`

Wolfsblvt replied:
Missed that one. Let me check functionality actually.
Thing is, jQuery's `.data()` call is not really setting a data attribute. That confused me at first too. It's not adding it to the actual DOM, so you can't query it via the usual `.dataset` on the JS level, and its not displayed in the DOM tree. This call just adds it in some kind of temporary memory, only accessible for jQuery itself, to save/link data to some actual DOM nodes.
I don't know why you'd actually use it that way.... but it was used in many places.

I just replaced all the actual HTML attributes, because thinking about HTML5 I am not a fan of making up HTML tags, if the actual `data-xxxx="value"` format exists for a good reason.

Wolfsblvt replied:
Changed it back. That one should stay like this. Didn't really wanna change **all** of the possible `data` references.

Wolfsblvt committed:
Add button to show connections for current char

Wolfsblvt committed:
Fix wrongly changed jQuery data "chid" attribute

Wolfsblvt committed:
Fix temporary persona toast

- Temporary persona toast was appearing when auto-lock was enabled, which didn't make sense

Wolfsblvt committed:
Highlight personas from current chat in popup

Wolfsblvt committed:
Add hint on how to use persona connections

Wolfsblvt committed:
Consistent headers in persona panel

Wolfsblvt committed:
Fix "Default" persona connection tooltip

Wolfsblvt committed:
Temporary persona toast respects setting

Wolfsblvt committed:
Remove all connections button for the popups

Cohee1207 committed:
Comment dupe default persona button

Cohee1207 committed:
Merge branch 'staging' into persona-improvements

Wolfsblvt committed:
Fix lock icon state on auto-lock to chat

Wolfsblvt committed:
Reduce persona toast spam

Wolfsblvt committed:
Don't look at this

Wolfsblvt committed:
Merge branch 'staging' into persona-improvements

Wolfsblvt committed:
Move persona buttons to right panel

- Move all buttons from the persona list to the right panel, to the already existing ones
- Common CSS class/style for '.buttons_block'
- Common CSS style for a red button
- Refactored a few persona buttons' functionalities

Wolfsblvt committed:
Add persona connection states to persona list

- Show persona connection/lock states in persona list (chat & char lock)
- Refactor persona images url to 'data-avatar-id' attribute
- Try make persona blocks' height consistent
- Fix persona list not correctly updating selected persona on navigating pages
- Fix group/char difference not correctly working
- Create common style template for hover-able images buttons/tags where the label only appears on hover

Wolfsblvt committed:
Streamline persona toasts, infos and states

- Added "info" block to persona description panel to show when a temporary persona is in use. Hide the long text behind a tooltip
- Reduce toast spam even further, not showing toasts when persona panel is open
- Restyle connection state buttons a bit
- Designed common 'info-block' utility to show markdown-like info blocks, with CSS styling

Wolfsblvt committed:
Add rename personas functionality

Cohee1207 committed:
Update persona upload methods to fetch

Wolfsblvt committed:
Merge branch 'staging' into persona-improvements

Cohee1207 committed:
Replace gray with opacity

Wolfsblvt committed:
Remove unused old username functions

Cohee1207 committed:
Align message block baseline

Wolfsblvt committed:
Merge branch 'persona-improvements' of https://github.com/SillyTavern/SillyTavern into persona-improvements

Wolfsblvt committed:
fix linting errors

Cohee1207 committed:
Rerender the list if migrated persona

Wolfsblvt committed:
Refactor: move showCharConnections

Cohee1207 committed:
Remove margin right from persona name

Cohee1207 committed:
Add missing personas on load

Cohee1207 committed:
Update rename tooltip

Wolfsblvt committed:
Refactor: move persona slash commands

Cohee1207 committed:
Clickable char personas

Wolfsblvt committed:
utility: let setInfoBlock also clear

Wolfsblvt committed:
Switch to char from persona panel decision tree

Wolfsblvt committed:
Update persona.js code documentation for exported

Wolfsblvt commented:
Should be ready now. I did all I could and wanted.

Wolfsblvt committed:
Merge branch 'staging' into persona-improvements

Wolfsblvt committed:
Fix persona UI states sometimes not updating

Cohee1207 committed:
Merge persona description counter with position header

Cohee1207 committed:
Don't choose chat persona if it's already chosen

Cohee1207 committed:
Navigate to current persona on switching chats

Cohee1207 committed:
Comment out persona preservation on character swap with personas open

Cohee1207 committed:
Only navigate to persona if switched by connection

Cohee1207 committed:
Merge branch 'staging' into persona-improvements
